zkevm-circuits/src/evm_circuit/execution/exp.rs

```diff
@@ -0,0 +1,301 @@
+            exponent: exponent_rlc,
+            exponentiation: exponentiation_rlc,
+            exponent_is_zero,
+            exponent_is_one,
+            single_step,
+            exponent_byte_size,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        _call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let [base, exponent, exponentiation] =
+            [step.rw_indices[0], step.rw_indices[1], step.rw_indices[2]]
+                .map(|idx| block.rws[idx].stack_value());
+
+        self.base.assign(region, offset, Some(base.to_le_bytes()))?;
+        self.exponent
+            .assign(region, offset, Some(exponent.to_le_bytes()))?;
+        self.exponentiation
+            .assign(region, offset, Some(exponentiation.to_le_bytes()))?;
+
+        let exponent_scalar = exponent
+            .to_scalar()
+            .expect("exponent should fit into scalar");
+        self.exponent_is_zero
+            .assign(region, offset, exponent_scalar)?;
+        self.exponent_is_one
+            .assign(region, offset, exponent_scalar, F::one())?;
+
+        let (base_sq, _) = base.overflowing_mul(base);
+        self.zero_rlc
+            .assign(region, offset, Some(U256::zero().to_le_bytes()))?;
+        self.base_sq
+            .assign(region, offset, Some(base_sq.to_le_bytes()))?;
+        self.base_sq_mul_gadget
+            .assign(region, offset, [base, base, U256::zero(), base_sq])?;
+        let single_step = exponent.eq(&U256::from(2u64));
+        self.single_step
+            .assign(region, offset, Value::known(F::from(single_step as u64)))?;
+
+        self.exponent_byte_size.assign(region, offset, exponent)?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use eth_types::{bytecode, Word};
+    use mock::TestContext;
+
+    use crate::test_util::run_test_circuits;
+
+    fn test_ok(base: Word, exponent: Word) {
+        let code = bytecode! {
+            PUSH32(exponent)
+            PUSH32(base)
+            EXP
+            STOP
+        };
+        assert_eq!(
+            run_test_circuits(
+                TestContext::<2, 1>::simple_ctx_with_bytecode(code).unwrap(),
+                None
+            ),
+            Ok(())
+        );
+    }
+
+    #[test]
+    fn exp_gadget_zero() {
+        test_ok(Word::zero(), Word::zero());
+        test_ok(Word::one(), Word::zero());
+        test_ok(0xcafeu64.into(), Word::zero());
+        test_ok(Word::MAX, Word::zero());
+    }
+
+    #[test]
+    fn exp_gadget_one() {
+        test_ok(Word::zero(), Word::one());
+        test_ok(Word::one(), Word::one());
+        test_ok(0xcafeu64.into(), Word::one());
+        test_ok(Word::MAX, Word::one());
+    }
+
+    #[test]
+    fn exp_gadget_simple() {
+        test_ok(2.into(), 5.into());
```

Comment:
this test will fail for the case `test_ok(2.into(), Word::from_str_radix("0x200000000000000000000000000000000", 16).unwrap());`.